test/common/network/listen_socket_impl_test.cc

```diff
@@ -2,43 +2,60 @@
 
 #include "common/network/utility.h"
 #include "common/network/listen_socket_impl.h"
+#include "test/test_common/network_utility.h"
 
 namespace Network {
 
-TEST(ListenSocket, All) {
-  // Test the case of a socket with given port and bind_to_port set to true.
-  TcpListenSocket socket1(uint32_t(15000), true);
+TEST(ListenSocket, BindIPv4Port) {
+  // Test the case of a socket with given address and port, and with bind_to_port set to true.
+  auto addr = Network::Test::findOrCheckFreePort("127.0.0.2:0", Address::SocketType::Stream);
+  EXPECT_LT(0U, addr->ip()->port());
+  TcpListenSocket socket1(addr, true);
   EXPECT_EQ(0, listen(socket1.fd(), 0));
-  EXPECT_EQ(15000U, socket1.localAddress()->ip()->port());
+  EXPECT_EQ(addr->ip()->port(), socket1.localAddress()->ip()->port());
+  EXPECT_EQ("127.0.0.2", socket1.localAddress()->ip()->addressAsString());
 
-  // Test the case of a socket with given tcp address and bind_to_port set to true.
-  TcpListenSocket socket2(Utility::resolveUrl("tcp://127.0.0.1:15002"), true);
-  EXPECT_EQ(0, listen(socket2.fd(), 0));
-  EXPECT_EQ("127.0.0.1:15002", socket2.localAddress()->asString());
+  // The address and port are bound already, should throw exception.
+  EXPECT_THROW(Network::TcpListenSocket socket2(addr, true), EnvoyException);
 
-  // The port is bound already, should throw exception.
-  EXPECT_THROW(Network::TcpListenSocket socket3(uint32_t(15000), true), EnvoyException);
+  // Test the case of a socket with fd and given address and port.
+  TcpListenSocket socket3(dup(socket1.fd()), addr);
+  EXPECT_EQ(addr->asString(), socket3.localAddress()->asString());
+}
 
-  // The address is bound already, should throw exception.
-  EXPECT_THROW(Network::TcpListenSocket socket4(Utility::resolveUrl("tcp://127.0.0.1:15002"), true),
-               EnvoyException);
+TEST(ListenSocket, BindIPv6Port) {
+  // Test the case of a socket with given address and port, and with bind_to_port set to true.
+  auto addr = Network::Test::findOrCheckFreePort("[::1]:0", Address::SocketType::Stream);
```

Comment:
There is only one IPv6 loopback address, so far as I know. Nonetheless, hidden now in a call to getSomeLoopbackAddress.